crates/wasm-host/src/context.rs

```diff
@@ -0,0 +1,1037 @@
+                local next_expiration = timestamp + expiration_window
+                if not expiration or timestamp > expiration then
+                    redis.call("set", expiration_key, next_expiration)
+                    redis.call("set", counter_key, 0)
+                    redis.call("expireat", expiration_key, next_expiration + 1)
+                    redis.call("expireat", counter_key, next_expiration + 1)
+                    expiration = next_expiration
+                end
+                local attempts = redis.call("incrby", counter_key, increment_delta)
+                return { attempts, expiration }
+                "#,
+            ),
+            check_rate_limit: redis::Script::new(
+                r#"
+                local counter_key = "bulwark:rl:" .. KEYS[1]
+                local expiration_key = counter_key .. ":exp"
+                local timestamp = tonumber(ARGV[1])
+                local attempts = tonumber(redis.call("get", counter_key))
+                local expiration = nil
+                if attempts then
+                    expiration = tonumber(redis.call("get", expiration_key))
+                    if not expiration or timestamp > expiration then
+                        attempts = nil
+                        expiration = nil
+                    end
+                end
+                return { attempts, expiration }
+                "#,
+            ),
+            increment_breaker: redis::Script::new(
+                r#"
+                local generation_key = "bulwark:bk:g:" .. KEYS[1]
+                local success_key = "bulwark:bk:s:" .. KEYS[1]
+                local failure_key = "bulwark:bk:f:" .. KEYS[1]
+                local consec_success_key = "bulwark:bk:cs:" .. KEYS[1]
+                local consec_failure_key = "bulwark:bk:cf:" .. KEYS[1]
+                local success_delta = tonumber(ARGV[1])
+                local failure_delta = tonumber(ARGV[2])
+                local expiration_window = tonumber(ARGV[3])
+                local timestamp = tonumber(ARGV[4])
+                local expiration = timestamp + expiration_window
+                local generation = redis.call("incrby", generation_key, 1)
+                local successes = 0
+                local failures = 0
+                local consec_successes = 0
+                local consec_failures = 0
+                if success_delta > 0 then
+                    successes = redis.call("incrby", success_key, success_delta)
+                    failures = tonumber(redis.call("get", failure_key)) or 0
+                    consec_successes = redis.call("incrby", consec_success_key, success_delta)
+                    redis.call("set", consec_failure_key, 0)
+                    consec_failures = 0
+                else
+                    successes = tonumber(redis.call("get", success_key))
+                    failures = redis.call("incrby", failure_key, failure_delta) or 0
```

Comment:
I believe this comes from Redis returning a non-numeric value for `get`s against a key that doesn't exist yet. If I'm remembering correctly, we need the `or 0` to ensure the variable only contains a numeric value.

Reply:
Sorry, I read the comment too quickly, that does seem odd. I added it everywhere. Based on what I saw just testing things out in `redis-cli`, I think it might only be needed for the `get` calls and not the `incrby` ones, but I don't think it hurts to just have it everywhere?